pkg/registry/core/service/storage/rest.go

```diff
@@ -541,33 +583,322 @@ func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableO
-		return r.secondaryServiceIPs
+	// if headless service then we bail out early (no clusterIPs management needed)
+	if len(oldService.Spec.ClusterIPs) > 0 && oldService.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		return nil, nil, nil
+	}
+
+	upgraded := len(oldService.Spec.IPFamilies) == 1 && len(service.Spec.IPFamilies) == 2
+	downgraded := len(oldService.Spec.IPFamilies) == 2 && len(service.Spec.IPFamilies) == 1
+
+	// CASE C:
+	if upgraded {
+		toAllocate := make(map[api.IPFamily]string)
+		// if secondary ip was named, just get it. if not add a marker
+		if len(service.Spec.ClusterIPs) < 2 {
+			service.Spec.ClusterIPs = append(service.Spec.ClusterIPs, "" /* marker */)
+		}
+
+		toAllocate[service.Spec.IPFamilies[1]] = service.Spec.ClusterIPs[1]
+
+		// allocate
+		allocated, err := rs.allocClusterIPs(service, toAllocate)
+		// set if successful
+		if err == nil {
+			service.Spec.ClusterIPs[1] = allocated[service.Spec.IPFamilies[1]]
+		}
+
+		return allocated, nil, err
 	}
 
-	return r.serviceIPs
+	// CASE D:
+	if downgraded {
+		toRelease := make(map[api.IPFamily]string)
+		toRelease[oldService.Spec.IPFamilies[1]] = oldService.Spec.ClusterIPs[1]
+		released, err := rs.releaseClusterIPs(toRelease)
+		return nil, released, err
+	}
+	// it was not an upgrade nor downgrade
+	return nil, nil, nil
 }
 
-func (r *REST) getAllocatorBySpec(service *api.Service) ipallocator.Interface {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) ||
-		service.Spec.IPFamily == nil ||
-		r.secondaryServiceIPs == nil {
-		return r.serviceIPs
+// releases allocated ClusterIPs for service that is about to be deleted
+func (rs *REST) releaseServiceClusterIPs(service *api.Service) (released map[api.IPFamily]string, err error) {
+	// external name don't get ClusterIPs
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		return nil, nil
 	}
 
-	secondaryAllocatorCIDR := r.secondaryServiceIPs.CIDR()
-	if (*(service.Spec.IPFamily) == api.IPv6Protocol) == netutil.IsIPv6CIDR(&secondaryAllocatorCIDR) {
-		return r.secondaryServiceIPs
+	// headless don't get ClusterIPs
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		return nil, nil
 	}
 
-	return r.serviceIPs
+	toRelease := make(map[api.IPFamily]string)
+	for _, ip := range service.Spec.ClusterIPs {
+		if netutil.IsIPv6String(ip) {
+			toRelease[api.IPv6Protocol] = ip
+		} else {
+			toRelease[api.IPv4Protocol] = ip
+		}
+	}
+	return rs.releaseClusterIPs(toRelease)
+}
+
+// attempts to default service ip families according to cluster configuration
+// while ensuring that provided families are configured on cluster.
+func (rs *REST) tryDefaultValidateServiceClusterIPFields(service *api.Service) error {
+	// can not do anything here
+	if service.Spec.Type == api.ServiceTypeExternalName {
+		return nil
+	}
+	// two families or two IPs with Single or PreferDualStack is not allowed
+	if service.Spec.IPFamilyPolicy != nil &&
```

Comment:
Validation is called twice. one before and one after the allocation logic run. If we put this step into validation then it will fail (when the allocator assign two families and two IPs). Having this step here (and not in validation) ensures that user input is validated (e.g. no two families/two IPs with Prefer..).